public_html/admin/model/sale/coupon.php

```diff
@@ -114,111 +133,168 @@ public function editCoupon($coupon_id, $data)
-            $this->db->query("UPDATE ".$this->db->table("coupons")." 
-							SET ".implode(',', $update)."
-							WHERE coupon_id = '".(int)$coupon_id."'");
+            $this->db->query(
+                "UPDATE ".$this->db->table("coupons")." 
+                SET ".implode(',', $update)."
+                WHERE coupon_id = '".(int) $coupon_id."'"
+            );
         }
 
         if (!empty($data['coupon_description'])) {
             foreach ($data['coupon_description'] as $language_id => $value) {
-                $update = array();
+                $update = [];
                 if (isset($value['name'])) {
                     $update["name"] = $value['name'];
                 }
                 if (isset($value['description'])) {
                     $update["description"] = $value['description'];
                 }
                 if (!empty($update)) {
-                    $this->language->replaceDescriptions('coupon_descriptions',
-                        array('coupon_id' => (int)$coupon_id),
-                        array(
-                            $language_id => array(
+                    $this->language->replaceDescriptions(
+                        'coupon_descriptions',
+                        ['coupon_id' => (int) $coupon_id],
+                        [
+                            $language_id => [
                                 'name'        => $value['name'],
                                 'description' => $value['description'],
-                            ),
-                        ));
+                            ],
+                        ]
+                    );
                 }
             }
         }
     }
 
     /**
-     * @param int   $coupon_id
+     * @param int $coupon_id
      * @param array $data
+     *
+     * @throws AException
      */
     public function editCouponProducts($coupon_id, $data)
     {
-        $this->db->query("DELETE FROM ".$this->db->table("coupons_products")." 
-						  WHERE coupon_id = '".(int)$coupon_id."'");
+        $this->db->query(
+            "DELETE FROM ".$this->db->table("coupons_products")." 
+            WHERE coupon_id = '".(int) $coupon_id."'"
+        );
         if (isset($data['coupon_product'])) {
             foreach ($data['coupon_product'] as $product_id) {
-                $this->db->query("INSERT INTO ".$this->db->table("coupons_products")." 
-									SET coupon_id = '".(int)$coupon_id."',
-										product_id = '".(int)$product_id."'");
+                $this->db->query(
+                    "INSERT INTO ".$this->db->table("coupons_products")." 
+                    SET coupon_id = '".(int) $coupon_id."',
+                        product_id = '".(int) $product_id."'"
+                );
+            }
+        }
+    }
+    /**
+     * @param int $coupon_id
+     * @param array $data
+     *
+     * @throws AException
+     */
+    public function editCouponCategories($coupon_id, $data)
+    {
+        $this->db->query(
```

Comment:
Need some validation before deleting all. If no $coupon_id, all records with 0 will be deleted. Bug

Also if you do validation of params in the beginning, there is no need to put (int) in every line. Too much